sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -514,6 +514,34 @@ case class OptimizeCodegen(conf: CatalystConf) extends Rule[LogicalPlan] {
 
 
 /**
+ * Reorders the predicates in `Filter` so more expensive expressions like UDF can evaluate later.
+ */
+object ReorderPredicatesInFilter extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f @ Filter(pred, child) =>
+      // Extracts deterministic suffix expressions from Filter predicate.
+      val expressions = splitConjunctivePredicates(pred)
+      // The beginning index of the deterministic suffix expressions.
+      var splitIndex = -1
+      (expressions.length - 1 to 0 by -1).foreach { idx =>
+        if (splitIndex == -1 && !expressions(idx).deterministic) {
+          splitIndex = idx + 1
+        }
+      }
+      if (splitIndex == expressions.length) {
+        // All expressions are non-deterministic, no reordering.
+        f
+      } else {
+        val (nonDeterminstics, deterministicExprs) = expressions.splitAt(splitIndex)
```

Comment:
Hmm, actually that's what I mean, probably some confusing with `non-deterministic` with `non-foldable`? I think we can skip them both in a short cut evaluation. as those expressions are not `stateful`(unfortunately, Spark SQL expression doesn't have the concept of `stateful`), so skip the evaluation of them are harmless, and this is exactly the short cut logic of expression `AND`.